Add columnar index mode overview (#7594)
elastic/docs-content: manage-data/data-store/columnar.md

---
navigation_title: Columnar storage
description: Learn when to use columnar index mode to store data once for analytics and search with a smaller storage footprint.
applies_to:
  stack: preview 9.5
  serverless: preview
products:
  - id: elasticsearch
---

# Columnar index mode [columnar-index-mode]

> Columnar index mode turns {{es}} into an analytical and search columnar store for the indices where you enable it. Instead of keeping multiple copies of each field for different query paths, columnar mode stores fields once as [doc values](elasticsearch://reference/elasticsearch/mapping-reference/doc-values.md) by default, which can reduce storage cost for high-volume, analytics-heavy data while keeping the same APIs, dashboards, and integrations.

Reviewer: ```suggestion
Columnar index mode turns {{es}} into an analytical and search columnar store for the indices where you enable it. Instead of keeping multiple copies of each field for different query paths, columnar mode stores fields once as [doc values](elasticsearch://reference/elasticsearch/mapping-reference/doc-values.md) by default. This helps reduce storage cost for high-volume, analytics-heavy data while keeping the same APIs, dashboards, and integrations.
```

Author: Thanks! I used this suggestion (with addition of "This strategy....") in e783c5d86c2f56fe2e1c80beec67871032876f14

> Columnar mode ships alongside existing index modes such as `standard`, [`logsdb`](/manage-data/data-store/data-streams/logs-data-stream.md), and [`time_series`](/manage-data/data-store/data-streams/time-series-data-stream-tsds.md).

Reviewer 2: * There's also the `[vectordb_document](https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/dense-vector#dense-vector-vectordb-document-mode)` mode as well. 
It's optimized for vector search use cases and applies settings and defaults tuned for indexing, merging, and searching dense vector data.

* And there's `lookup` which is a mode that can be used for [LOOKUP JOIN](https://www.elastic.co/docs/reference/query-languages/esql/esql-lookup-join) in ES|QL.

I can see you're linking to the [Index mode settings](https://www.elastic.co/docs/reference/elasticsearch/index-settings/index-modules#index-mode-setting) page from your doc, so that's where the info I'm basing this comment comes from.

Reviewer 2: This makes me think we need to add more information in this section about these as well (as a separate task).

For now, we can at least mention these and link to their respective reference docs. Wdyt?

Reviewer 3: I think that the ones mentioned here should suffice for now, I would only add `vectordb_document` if you want more coverage. Lookup indices are pretty special and not at the same level as the other ones. 

We should revisit at some point the way we talk about all those index modes, move all of them under a new hierarchy and talk about them as (storage) engines Elasticsearch provides. But we can discuss about this as a follow-up.

Author: Thanks, I'll leave as-is for now but agree we should follow-up in another issue.

> You choose it per index (or in a template) at creation time; you can't change the mode after the index exists.

Reviewer 3: ```suggestion
You choose it per index (or in a template) at creation time; you can't change the mode after the index has been created.
```

This page explains what columnar index mode is, when to use it, and how it fits with the rest of the {{es}} data store.
<!--
TO-DO after https://github.com/elastic/elasticsearch/pull/151739 merges
For enablement steps, sorting, `_source` modes, and limitations, refer to [Columnar index mode](elasticsearch://reference/columnar/index.md) in the {{es}} reference.
-->

## When to use columnar index mode [when-to-use-columnar]

Choose columnar index mode when your data is written in volume, queried analytically (filters, aggregations, and dashboards), and retained for a long time. Typical fits include:

> * Logs and observability at scale: High ingest rates where storage and aggregation cost dominate, and you still need full-text search on message fields.

Reviewer 3: ```suggestion
* Logs, traces and observability at scale: High ingest rates where storage and aggregation cost dominate, and you still need full-text search on message fields.
```

* Security telemetry and threat hunting: Faceted exploration and historical lookups over large event stores, without giving up search behavior for pivot and lookup workflows.
* Operational and business analytics: Dashboards and aggregations over application events, transactions, or IoT readings that would otherwise push you to a separate analytics store.

Columnar mode isn't a universal default.

> Prefer a standard index (or another specialized mode) when your workload is document-centric search that needs the original JSON `_source` as submitted, or when you rely on inverted indexes and related structures on most fields by default.

Reviewer: ```suggestion

Prefer a `standard` index (or another specialized mode) when your workload is document-centric search that needs the original JSON `_source` as submitted, or when you rely on inverted indexes and related structures on most fields by default.
```

Reviewer: Adding line break and putting "standard" in backticks to match the format used earlier on the page

Reviewer 2: Should `indexes` be `indices`? I've mostly seen the pl form of _indices_ in our docs

Reviewer 3: We try to use indices as the plural of index in general. Indexes is not wrong, but it is less used in the industry

For metrics that need time series dimensions and metric field semantics, use a [time series data stream](/manage-data/data-store/data-streams/time-series-data-stream-tsds.md) instead.

> For logs that should keep the current `logsdb` defaults without switching to a fully columnar store, use a [logs data stream](/manage-data/data-store/data-streams/logs-data-stream.md).

Reviewer: I wonder if all this info about use cases would work better in a table. Might be easier to reference which mode to use when

Author: I think it's actually better covered at a higher level, so I'll defer that change to https://github.com/elastic/docs-content/issues/7617

## Columnar modes [columnar-modes]

Two values of [`index.mode`](elasticsearch://reference/elasticsearch/index-settings/index-modules.md#index-mode-setting) enable columnar storage:

`columnar`
:   A general-purpose columnar store with no use-case-specific defaults.

> :   Use it for bare indices and data streams that aren't log-oriented.

Reviewer 3: Please my question on the other PR on this

https://github.com/elastic/elasticsearch/pull/151739/changes#r3668414160

Depending on the answer by the team, we may need to update this statement here as well

Reviewer 3: This is OK and resolved - mentioning data streams is correct

Author: Thanks for confirming!

`logsdb_columnar`
:   The same columnar storage behavior with logging-oriented defaults, including a default `@timestamp` mapping and index sorting when `@timestamp` and `host.name` are present.
:   Use it for log data when you want columnar storage rather than the default [`logsdb`](/manage-data/data-store/data-streams/logs-data-stream.md) mode.

> Both modes apply to indices and to data stream backing indices configured through templates.

Reviewer 3: Same as above: https://github.com/elastic/elasticsearch/pull/151739/changes#r3668414160

## How it works [how-columnar-works]

At a high level, columnar index mode changes storage defaults while leaving the document and query model familiar:

* Store once, by field: Non-text fields are stored as doc values and aren't indexed by default, which avoids paying for structures your analytics workload may not need.
* Keep search where it matters: Text fields remain indexed by default so full-text search continues to work on fields such as log messages.

> * Reconstruct `_source` when needed: The original JSON document body isn't stored as a separate copy. {{es}} returns a reconstructed or columnar `_source` representation at query time (or can store a columnar `_source` for faster retrieval).

Reviewer 3: We should mention that not storing the columnar source is available only for Enterprise licenses. It is stored by default on basic and can not be turned off.  On Enterprise license tier it is not stored by default, with an option to store it for faster retrieval

Author: Since that's covered in detail in the other page (and I think maybe a little too low-level for this overview), I've removed it in e783c5d86c2f56fe2e1c80beec67871032876f14. Thanks for pointing out the complexity!
